ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/CommandCompletionsTable.scala

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2020 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.platform.store.dao
+
+import anorm.{Row, RowParser, SimpleSql, SqlParser, SqlStringInterpolation, ~}
+import com.digitalasset.daml.lf.data.Ref
+import com.digitalasset.ledger.ApplicationId
+import com.digitalasset.ledger.api.v1.command_completion_service.CompletionStreamResponse
+import com.digitalasset.ledger.api.v1.completion.Completion
+import com.digitalasset.platform.store.CompletionFromTransaction.{toApiCheckpoint, toErrorCode}
+import com.digitalasset.platform.store.entries.LedgerEntry
+import com.google.rpc.status.Status
+
+object CommandCompletionsTable {
+
+  import SqlParser.{date, int, long, str}
+
+  private val acceptedCommandParser: RowParser[CompletionStreamResponse] =
+    long("completion_offset") ~ date("record_time") ~ str("command_id") ~ str("transaction_id") map {
+      case offset ~ recordTime ~ commandId ~ transactionId =>
+        CompletionStreamResponse(
+          checkpoint = toApiCheckpoint(recordTime.toInstant, offset),
+          completions = Seq(Completion(commandId, Some(Status()), transactionId)))
+    }
+
+  private val rejectedCommandParser: RowParser[CompletionStreamResponse] =
+    long("completion_offset") ~ date("record_time") ~ str("command_id") ~ int("status_code") ~ str(
+      "status_message") map {
+      case offset ~ recordTime ~ commandId ~ statusCode ~ statusMessage =>
+        CompletionStreamResponse(
+          checkpoint = toApiCheckpoint(recordTime.toInstant, offset),
+          completions = Seq(Completion(commandId, Some(Status(statusCode, statusMessage)))))
+    }
+
+  private val checkpointParser: RowParser[CompletionStreamResponse] =
+    long("completion_offset") ~ date("record_time") map {
+      case offset ~ recordTime =>
+        CompletionStreamResponse(
+          checkpoint = toApiCheckpoint(recordTime.toInstant, offset),
+          completions = Seq())
+    }
+
+  val parser: RowParser[CompletionStreamResponse] =
+    acceptedCommandParser | rejectedCommandParser | checkpointParser
```

Comment:
How does anorm decide when to choose a different alternative? Because the row in the result set contains all columns. I suppose `str("transaction_id")` only matches if it's the string is not null?

Reply:
Yes, exactly. :slightly_smiling_face: